dialogue-core/src/main/java/com/palantir/dialogue/core/StickyRoutingChannel.java

```diff
@@ -0,0 +1,107 @@
+/*
+ * (c) Copyright 2021 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.dialogue.core;
+
+import com.github.benmanes.caffeine.cache.CacheLoader;
+import com.github.benmanes.caffeine.cache.Caffeine;
+import com.github.benmanes.caffeine.cache.LoadingCache;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.palantir.dialogue.Channel;
+import com.palantir.dialogue.Endpoint;
+import com.palantir.dialogue.Request;
+import com.palantir.dialogue.Response;
+import com.palantir.dialogue.RoutingAttachments;
+import com.palantir.dialogue.RoutingAttachments.RoutingKey;
+import java.time.Duration;
+import javax.annotation.CheckForNull;
+import org.immutables.value.Value;
+
+@Value.Enclosing
+final class StickyRoutingChannel implements Channel {
+
+    private final Channel globalQueue;
+    private final StickyQueues stickyQueues;
+
+    StickyRoutingChannel(Config cf, Channel globalQueue, LimitedChannel nodeSelectionChannel) {
+        this.globalQueue = globalQueue;
+        this.stickyQueues = new StickyQueues(cf, nodeSelectionChannel);
+    }
+
+    @Override
+    public ListenableFuture<Response> execute(Endpoint endpoint, Request request) {
+        RoutingKey routingKey = request.attachments().getOrDefault(RoutingAttachments.ROUTING_KEY, null);
+        if (routingKey == null) {
+            return globalQueue.execute(endpoint, request);
+        } else {
+            return stickyQueues.execute(ImmutableStickyRoutingChannel.QueueKey.of(routingKey), endpoint, request);
+        }
+    }
+
+    private static final class StickyQueues {
+
+        private final LoadingCache<QueueKey, ExpiringQueue> queues;
```

Comment:
I think this cache can be avoided if the routing key is/has the sticky session queue. That way work is front-loaded on sticky session creation, lifecycle is well bounded to the scope of the sticky channel reference, and we don't have to worry about half of the hardest problems in CS.

Reply:
Uuuu, I think that would be nice, lemme see if that works.

Reply:
Have a look, I've done it this way.